src/sources/generic.rs

```diff
@@ -40,51 +40,77 @@
 //! If you need to directly work with a [`RawFd`](std::os::unix::io::RawFd), rather than an
 //! FD-backed object, see [`Generic::from_fd`](Generic#method.from_fd).
 
-use std::io;
-use std::os::unix::io::AsRawFd;
+use std::{marker::PhantomData, os::unix::io::AsRawFd};
 
 use crate::{EventSource, Interest, Mode, Poll, PostAction, Readiness, Token, TokenFactory};
 
 /// A generic event source wrapping a FD-backed type
 #[derive(Debug)]
-pub struct Generic<F: AsRawFd> {
+pub struct Generic<F: AsRawFd, E = std::io::Error> {
     /// The wrapped FD-backed type
     pub file: F,
     /// The programmed interest
     pub interest: Interest,
     /// The programmed mode
     pub mode: Mode,
     token: Box<Token>,
+
+    /// This exists only to allow us the make the associated types `Ret` and
+    /// `Error` be generic. It should be revisited when *[generic associated
+    /// types]* are complete and stable.
+    ///
+    /// [generic associated types]: https://github.com/rust-lang/rfcs/blob/master/text/1598-generic_associated_types.md
```

Comment:
What kind of future revision are you thinking about?

Reply:
I think, but I'm not 100% sure, that when GATs are stable you should be able to get rid of the phantom data marker and just have the `<E>` in the `impl<F: AsRawFd, E: Into<Box<dyn Error etc>>> EventSource for Generic<F>` instead.

Reply:
I don't think that will work, using GAT here would require changing the `EventSource` trait itself, and I'm not sure it'd be appropriate.

Reply:
Okay I re-read the RFC and blog post and you're right, I'll remove this.